### PR TITLE
Encrypted SNI

### PR DESCRIPTION
**Pull request by kazuho**
implements https://datatracker.ietf.org/doc/draft-rescorla-tls-esni/

ToDo:
* [x] refactor the API related to private keys
  * ptls_parse_esni should accept a callback that is used to load the private key (which could return an `ptls_key_exchange_context_t *` encapsulating a private key protected by neverbleed)
* [x] add tests

### CODE REVIEW

File: src/esni.c [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+            ptls_base64_encode(esni_keys.base, esni_keys.len, esni_keys_base64);
+            printf("  TXT record: \"%s\" (%zu bytes)\n", esni_keys_base64, strlen(esni_keys_base64));
+            free(esni_keys_base64);
+            ptls_esni_dispose(&esni);
+            src = end;
+        });
+    } while (src != end);
+
+    ret = 0;
+Exit:
```
Comment (Lekensteyn) [inline src/esni.c]: Running `picotls-esni -r` reports a leak, fix it by adding this line here:

        ptls_buffer_dispose(&buf);

File: t/cli.c [hunk 2]
```diff
@@ -265,6 +266,8 @@ static int run_client(struct sockaddr *sa, socklen_t salen, ptls_context_t *ctx,
 {
     int fd;
 
+    hsprop->client.esni_keys = resolve_esni_keys(server_name);
```
Comment (Lekensteyn) [inline t/cli.c]: The returned iovec is never freed:
```
Direct leak of 1024 byte(s) in 1 object(s) allocated from:
    #0 0x7f1b27319d99 in __interceptor_malloc /build/gcc/src/gcc/libsanitizer/asan/asan_malloc_linux.cc:86
    #1 0x555e50d2c49f in ptls_buffer_reserve lib/picotls.c:481
    #2 0x555e50d2c5b0 in ptls_buffer__do_pushv lib/picotls.c:499
    #3 0x555e50d2175c in ptls_base64_decode lib/pembase64.c:189
    #4 0x555e50d1de88 in resolve_esni_keys t/util.h:331
    #5 0x555e50d1de88 in run_client t/cli.c:269
    #6 0x555e50d1de88 in main t/cli.c:455
    #7 0x7f1b2638e06a in __libc_start_main (/usr/lib/libc.so.6+0x2306a)
    #8 0x555e50d1e419 in _start (/tmp/pico/cli+0xc419)
```

File: src/esni.c (hunk 1 above)
Comment (kazuho) [reply src/esni.c]: We no longer have the `-r` option :-)

File: t/cli.c (hunk 2 above)
Comment (kazuho) [reply t/cli.c]: Fixed in cbabf29.

**Comment by kazuho** [comment]
The PR is ready, will soon be merged to master.

**Pull request merged**